src/llm/servable_initializer.cpp

```diff
@@ -117,36 +117,107 @@ void GenAiServableInitializer::loadTextProcessor(std::shared_ptr<GenAiServablePr
         return StatusCode::LLM_NODE_DIRECTORY_DOES_NOT_EXIST;
     }
 
     return StatusCode::OK;
 }
 
+Status determinePipelineType(PipelineType& pipelineType, const mediapipe::LLMCalculatorOptions& nodeOptions, const std::string& graphPath) {
+    // Assuming that models_path is always set
+    std::string parsedModelsPath;
+    auto status = parseModelsPath(parsedModelsPath, nodeOptions.models_path(), graphPath);
+    if (status != StatusCode::OK) {
+        return status;
+    }
+
+    std::filesystem::path parsedModelsPathFs(parsedModelsPath);
+    // Existence of embeddings models indicates we are dealing with VLM pipeline
+    bool isVLM = (std::filesystem::exists(parsedModelsPathFs / "openvino_text_embeddings_model.xml") &&
+                  std::filesystem::exists(parsedModelsPathFs / "openvino_vision_embeddings_model.bin"));
+
+    // If pipeline type is not explicitly defined by the user, we need to determine it based on the content of the models directory and configuration
+    if (nodeOptions.pipeline_type() == mediapipe::LLMCalculatorOptions::UNDEFINED_PIPELINE_TYPE) {
+        if (nodeOptions.device() == "NPU") {
+            if (isVLM) {
+                pipelineType = PipelineType::VLM;
+            } else {
+                pipelineType = PipelineType::TEXT;
+            }
+        } else {
+            if (isVLM) {
+                pipelineType = PipelineType::VLM_CB;
+            } else {
+                pipelineType = PipelineType::TEXT_CB;
+            }
+        }
+    } else {
+        switch (nodeOptions.pipeline_type()) {
+        case mediapipe::LLMCalculatorOptions::TEXT:
+            pipelineType = PipelineType::TEXT;
+            break;
+        case mediapipe::LLMCalculatorOptions::VLM:
+            pipelineType = PipelineType::VLM;
+            break;
+        case mediapipe::LLMCalculatorOptions::TEXT_CB:
+            pipelineType = PipelineType::TEXT_CB;
+            break;
+        case mediapipe::LLMCalculatorOptions::VLM_CB:
+            pipelineType = PipelineType::VLM_CB;
+            break;
+        case mediapipe::LLMCalculatorOptions::CONTINUOUS_BATCHING:
+            pipelineType = PipelineType::TEXT_CB;
+            break;
+        case mediapipe::LLMCalculatorOptions::VISUAL_LANGUAGE_MODEL:
+            pipelineType = PipelineType::VLM_CB;
+            break;
+        default:
+            SPDLOG_LOGGER_ERROR(modelmanager_logger, "LLM node options do not contain any recognized pipeline configuration.");
+            return StatusCode::INTERNAL_ERROR;
+        }
+
+        if (isVLM && (pipelineType == PipelineType::TEXT || pipelineType == PipelineType::TEXT_CB)) {
+            SPDLOG_LOGGER_ERROR(modelmanager_logger, "Models directory content indicates VLM pipeline, but pipeline type is set to non-VLM type.");
+            return StatusCode::INTERNAL_ERROR;
+        }
+
+        if (!isVLM && (pipelineType == PipelineType::VLM || pipelineType == PipelineType::VLM_CB)) {
```

Comment:
```suggestion
        if (!isVLM && (pipelineType != PipelineType::TEXT && pipelineType != PipelineType::TEXT_CB)) {
```

isnt it safer to manually add it here everytime we need?

Reply:
Could you elaborate why? Current condition is more readable in my opinion as it's limited to VLM "terminology" (easy to associate isVLM with comparing against VLM pipeline types). Not sure I see the benefit of switching here.

Reply:
Readable yes, but in case we add new kind of pipeline, (and we forget to include it here) the if will return true, but it shouldnt right?